### PR TITLE
Adding allowShardReordering option to SearchRequest

### PR DESCRIPTION
**Pull request by pmpailis**
Spin off from this [PR comment](https://github.com/elastic/elasticsearch/pull/112648/files#r1755156315). 

In the above PR, we'd like to explicitly disable shard reordering for all search requests that may happen within a PIT to ensure consistent and comparable `shardIndex` values across results. In this PR, we add a new `allowShardReordering` option to `SearchRequest` which defaults to `true`, and will be used by the comment snippet in the referenced PR.

### CODE REVIEW

**Comment by javanna** [comment]
Is it an option to instead disable the can match phase for this scenario that you have in mind? Or is it a case of "we need/want can match, but we don't want it to resort the shards?" . Can you share more of the high-level purpose of this change? I don't think I follow the "ensure consistent and comparable shardIndex values across results". Intuitively, I thought that sorting the shards consistently provides for that?

**Comment by pmpailis** [comment]
Thanks for taking a look @javanna ! The idea is that since the RRF retriever executes >= 2 queries independently  through `msearch` and then tries to merge and rank their results, we want to be able to properly identify and correlate them for computing the final RRF ranking scores. To do that we rely on a [RankKey](https://github.com/elastic/elasticsearch/blob/ce1c55f5381072ea3fd142a5af2466bdf764856a/server/src/main/java/org/elasticsearch/search/rank/RankDoc.java#L46) that is a record of `<docId, shardIndex>`. 

So, in order to have comparable results across multiple searches (one could be a range query, another a knn and so on) we would need to make sure that `shardIndex` 0 always refers to the same `shard` , hence why we wanted to disable sorting. 

But looking more into it, even if we disable sorting and still execute the can_match phase,  we could still end up with different shardIndex references, since IIUC, in `AbstractSearchAsyncAction`  the `shardIndex` is computed based only on the shards actually contributing to search, and we ignore (in the context of `shardIndex`) the ones that can be skipped.  So we can start by just disabling can-match completely and might revisit to add this optimization once we have a better way to identify docs & shards. 

TLDR; closing this PR and might revisit in the future :) Thanks again for taking a look @javanna !
